### DISCUSSION TITLE
TvOS support

### DISCUSSION THREAD

**Discussion by henninghall** [wix/Detox#2853] [Ideas]
It would be great to add TvOS support. 

I have tried to run detox with a TvOS build on an Apple TV simulator. Initially, detox doesn't manage to open the app. Could it be because of a click event instead of a "select"-press? If I help detox to open the app it gets stuck on the first test with the error message:

`Simulator: Terminating capture support for coordinator '<DeviceCoordinat`
`ERROR: [APP_UNREACHABLE] Failed to reach the app over the web socket connection.`

**Comment by TrebuhD** [comment]
This was discussed in 2018 (https://github.com/wix/Detox/issues/836) and @LeoNatan 's conclusion was that it was not possible due to Detox's underlying framework (EarlGrey) not supporting tvOS. 

I don't know if anything's changed since then, but just dropping here to say this would be really great to have!

**Reply by henninghall** [reply]
Allright, I looked into it and tvOS seems to have been supported by EarlGrey since 2020, so hopefully this could be possible after all 👍 
https://github.com/google/EarlGrey/pull/1202

**Reply by d4vidi** [reply]
Not to say that this solves it for TvOS, but in Detox v18 we've dropped any usage of EarlGrey, in favor of our own home-brewed, performance-boosted solution: [DetoxSync](https://github.com/wix/DetoxSync).

Unfortunately, we don't have the capacity for looking into this ATM. However, seems like #836 might be out of date.

**Comment by SebastienChauvin** [comment]
Hello, 

I had a quick try today with detox on our react-native-tvos app.

Out of the box does not work for several reasons:
- frameworks are iOS only so we have a silent error where the test runner is never connected to the app as the app does not have the detox framework built in.
- when activating the framework for tvOS: objective C headers and Swift code referring to iOS only code
  
But we are not far off !

I am wondering if anyone from the detox time would have any idea / suggestion on the best ways to have a compatible framework. Detox looks good to us and we are considering to spend some time on a PR to have tvOS support.

How I got a POC to run:
- used detox from src with https://wix.github.io/Detox/docs/introduction/debugging/#setting-detox-up-as-a-compiling-dependency
- activated tvos build on my project
- removed everything that did not compile : [tvos_poc.patch.gz](https://github.com/wix/Detox/files/10320219/tvos_poc.patch.gz)
  - UISwitch, UIPicker, UISlider
  - WebKit
  - orientation
  - COSTouchVisualizer
  - UserNotification
  - UIColor systemBackgroundColor

**Reply by noomorph** [reply]
@d4vidi @asafkorem  Please don't miss this post 🚨

**Reply by asafkorem** [reply]
Hi @SebastienChauvin,

I'm sorry, I don't have any experience or familiarity with any UI testing framework for react-native-TvOS apps. However, it's possible that Detox may offer support for TvOS apps (in a third-party way or built-in) with some additional work. This will involve creating a new native client for tvOS apps and making some changes on the JavaScript side. Here is an article that might be helpful: https://wix.github.io/Detox/docs/articles/third-party-drivers/#writing-a-new-third-party-driver.

**Reply by asafkorem** [reply]
Unfortunately, we don't have plans to focus on this in the near future, so it may be necessary for the community to take the lead on adding support for TvOS apps. 

One option to consider is working on a third-party driver that can be integrated into the project later on. I would be happy to offer guidance and assistance with the project and its architecture if you decide to pursue this. 

Alternatively, you could consider creating a local copy of Detox and making the necessary adjustments (as you did), but this would mean that you wouldn't be able to receive future updates to the main Detox project.

**Reply by asafkorem** [reply]
Another option to consider is contributing to Detox directly. This would involve working with our team to review and integrate your changes into the main project, and may take more time (as it requires more of our involvement).

If you decide to move forward with adding support for TvOS apps in Detox, I'd be happy to have a direct channel of communication to discuss the various alternatives and the steps involved. You can find me on our Discord at https://discord.gg/CkD5QKheF5.

**Reply by SebastienChauvin** [reply]
Perfect thanks a bunch for your answer @asafkorem.

We will discuss our plans and priorities with the rest of the team and come back to this topic within a few days.
